### ISSUE TITLE
Training is unstable and val_loss is Nan

### ISSUE THREAD

**Issue by guishilike** [david8862/keras-YOLOv3-model-set#152]
训练极不稳定，loss正常情况，但是val_loss容易出现Nan。

**Comment by david8862** [comment]
@guishilike 可以提供一下你的训练配置以及数据集的大致情况么

**Comment by guishilike** [comment]
首先我尝试的是一个单类检测，数据集数量不是很多，自己手动标注，质量不高。
去除了部分数据增强（有些数据增强并不友好，会导致置信度很低）
在batch size为1到2 的情况下几乎无法训练，几轮后val loss便会出现nan。
提高batch size到8，情况略有改善，小概率训练成功，测试效果还行，但是再往上提高batch size所需显存就不够了。
我不否定数据集质量带来的影响。起初我认为是数据质量低导致的，但是最近看到很多其他仓库评论也有类似的说法，从反馈来看模型主干应该是没有问题，有没有可能在损失函数那存在精度丢失情况，导致后续运算出现nan。
因为同样的数据在一台mac系统上训练没有出现nan（虽然实验的次数不多)。

**Comment by david8862** [comment]
单类检测的话，或许可以参考YOLOv5 repo中的做法，在loss计算中去掉class_loss部分，只考虑box_loss和confidence_loss

**Comment by guishilike** [comment]
谢谢，我会参考的。
实际我想尝试的是多类，目前只是查看该类识别的效果。
十分抱歉，没有时间去细看这些细节。
训练过程主要是location_loss异常，采用的diou,我觉得正常情况下不应该出现nan。

**Comment by guishilike** [comment]
就类似于这样
`29/29 [==============================] - ETA: 0s - loss: 4538.6938 - location_loss: 7.0715 - confidence_loss: 4487.7080 - class_loss: 27.3298 
Epoch 00001: val_loss did not improve from inf`
`29/29 [==============================] - 405s 14s/step - loss: 4538.6938 - location_loss: 7.0715 - confidence_loss: 4487.7080 - class_loss: 27.3298 - val_loss: nan - val_location_loss: nan - val_confidence_loss: 1094371.3750 - val_class_loss: 58642.4453
Epoch 2/200`
`D:\anaconda3\envs\tfenv\lib\site-packages\tensorflow\python\keras\callbacks.py:1291: RuntimeWarning: invalid value encountered in less
  if self.monitor_op(current, self.best):`
`D:\anaconda3\envs\tfenv\lib\site-packages\tensorflow\python\keras\callbacks.py:2347: RuntimeWarning: invalid value encountered in less
  self.monitor_op = lambda a, b: np.less(a, b - self.min_delta)`
`D:\anaconda3\envs\tfenv\lib\site-packages\tensorflow\python\keras\callbacks.py:1664: RuntimeWarning: invalid value encountered in less
  if self.monitor_op(current - self.min_delta, self.best):`
`29/29 [==============================] - ETA: 0s - loss: 498.6384 - location_loss: 5.9935 - confidence_loss: 461.1933 - class_loss: 14.5822 
Epoch 00002: val_loss did not improve from inf`
`29/29 [==============================] - 401s 14s/step - loss: 498.6384 - location_loss: 5.9935 - confidence_loss: 461.1933 - class_loss: 14.5822 - val_loss: nan - val_location_loss: nan - val_confidence_loss: 20884.1465 - val_class_loss: 846.5333
Epoch 3/200`

**Comment by david8862** [comment]
DIOU loss因为收敛速度较快，对于数据集规模较小的任务可能更容易过拟合。或许你可以试试退回到YOLOv3的标准location_loss，相对比较稳定

**Comment by guishilike** [comment]
即使是过拟合，不应该是值偏大吗，不应该导致nan

**Comment by david8862** [comment]
@guishilike DIOU loss的计算图中针对被0除等常见的数值溢出情形是做了保护的，所以正常情况下应该不会出现nan的结果。你可以尝试将计算过程中的一些tensor导入到tensorboard里跟踪，看看是哪里发生了溢出
